jinja2/filters.py

```diff
@@ -940,6 +940,37 @@ def do_attr(environment, obj, name):
     return environment.undefined(obj=obj, name=name)
 
 
+@contextfilter
+def do_applymacro(*args, **kwargs):
+    """Applies a macro by reference instead of directly.
+
+    Calling a macro directly like,
+
+    {{ event_detail(event) }}
+
+    Is the equivalent with applymacro,
+
+    {{ event|applymacro(macroname='event_detail') }}
+
+    or
+
+    {{ event|applymacro(macro=event_detail) }}
+
+    Which then allows you to use the macro in other things like map,
+
+    {{ events|map('applymacro', macro=event_detail)|join(', ') }}
+
+    """
+    context = args[0]
+    args = args[1:]
+
+    macro = kwargs.pop('macro', None)
```

Comment:
Let's have one explicit arg and switch on it's type automatically.

Reply:
Do we even need both options? What's the benefit of a macro name instead of passing a macro? If you want to pick the macro dynamically in the code you can always use `{% set macro = foo if x else bar %}` with foo/bar being macros.

Reply:
Can the actual macro name come from the outside (as the template param)? I suppose, yes.

Reply:
Do you mean one explicit arg, as in `macro` as opposed to a change to the signature of the function, `def do_map(macro, *args, **kwargs)`?

> switch on it's type automatically

I.E. check if it's a `str` and if so look for it in `context.vars`?

Reply:
If we allow strings, we need to do,

    macro = kwargs.pop('macro')
    if isinstance(macro, str):
        if '.' in macro:
            templatemod, _, mname = macro.partition('.')
            macro = getattr(context.vars[templatemod], mname)

        else:
            macro = context.vars[macro]

So let strings be passed, or just allow the macro to be passed in directly, and not mess around with that?

Reply:
Without feedback, I'm marking this as resolved.  If you guys want this removed just say, and I'll pull it, but if you're okay with it, we'll leave it as is.